app/controllers/api/v2/notifications_controller.rb

```diff
@@ -80,7 +80,19 @@ def load_grouped_notifications
     return [] if @notifications.empty?
 
     MastodonOTELTracer.in_span('Api::V2::NotificationsController#load_grouped_notifications') do
-      NotificationGroup.from_notifications(@notifications, pagination_range: (@notifications.last.id)..(@notifications.first.id), grouped_types: params[:grouped_types])
+      pagination_range = (@notifications.last.id)..@notifications.first.id
+
+      if @notifications.size < limit_param(DEFAULT_NOTIFICATIONS_LIMIT)
+        if params[:min_id].present?
+          pagination_range = @notifications.last.id...(params[:max_id]&.to_i)
+        else
+          range_start = params[:since_id]&.to_i
+          range_start += 1 unless range_start.nil?
+          pagination_range = range_start..(@notifications.first.id)
+        end
+      end
```

Comment:
I found this to be really hard to follow. A little thing that might have helped me would have been to extract the conditions into helper methods that are named in a way that makes it clearer what cases these are.

So something like:

```ruby
if last_page?
  if paginating_down?
    # ...
  elsif paginating_up?
    # ...
  end
end
```

(I am not totally sure the naming makes sense, but I hope you get my meaning.)